Kernel/FileSystem/FATFileSystem.h

```diff
@@ -0,0 +1,177 @@
+    u8 drive_number;
+    u8 unused4;
+    u8 signature;
+    u32 volume_id;
+    char volume_label_string[11];
+    char system_identifier_string[8];
+};
+static_assert(sizeof(FAT32BootRecord) == 90);
+
+enum class FATAttributes : u8 {
+    ReadOnly = 0x01,
+    Hidden = 0x02,
+    System = 0x04,
+    VolumeID = 0x08,
+    Directory = 0x10,
+    Archive = 0x20,
+    LongFileName = 0x0F
+};
+
+AK_ENUM_BITWISE_OPERATORS(FATAttributes);
+
+struct [[gnu::packed]] FATEntry {
+    char filename[8];
+    char extension[3];
+    FATAttributes attributes;
+    u8 unused1;
+    u8 creation_time_seconds;
+    u16 creation_time;
+    u16 creation_date;
+    u16 last_accessed_date;
+    u16 first_cluster_high;
+    u16 modification_time;
+    u16 modification_date;
+    u16 first_cluster_low;
+    u32 file_size;
+};
+static_assert(sizeof(FATEntry) == 32);
+
+struct [[gnu::packed]] FATLongFileNameEntry {
+    u8 entry_index;
+    u16 characters1[5];
+    FATAttributes attributes;
+    u8 entry_type;
+    u8 checksum;
+    u16 characters2[6];
+    u16 zero;
+    u16 characters3[2];
+};
+static_assert(sizeof(FATLongFileNameEntry) == 32);
+
+class FATInode;
+
+class FATFS final : public BlockBasedFileSystem {
+    friend FATInode;
+
+public:
+    static ErrorOr<NonnullLockRefPtr<FileSystem>> try_create(OpenFileDescription&);
+
+    virtual ~FATFS() override = default;
+    virtual ErrorOr<void> initialize() override;
+    virtual StringView class_name() const override { return "FATFS"sv; }
+    virtual Inode& root_inode() override;
+
+private:
+    FATFS(OpenFileDescription&);
+
+    static constexpr u8 signature_1 = 0x28;
+    static constexpr u8 signature_2 = 0x29;
+
+    static constexpr u32 first_data_cluster = 2;
+
+    FAT32BootRecord const* boot_record() const { return reinterpret_cast<FAT32BootRecord const*>(m_boot_record->data()); };
+
+    BlockBasedFileSystem::BlockIndex first_block_of_cluster(u32 cluster) const;
+
+    OwnPtr<KBuffer> m_boot_record;
+    LockRefPtr<FATInode> m_root_inode;
+    u32 m_first_data_sector;
+};
+
+class FATInode final : public Inode {
+    friend FATFS;
+
+public:
+    virtual ~FATInode() override = default;
+
+    static ErrorOr<NonnullLockRefPtr<FATInode>> create(FATFS&, FATEntry, Vector<FATLongFileNameEntry> const& = {});
+
+    FATFS& fs() { return static_cast<FATFS&>(Inode::fs()); }
+    FATFS const& fs() const { return static_cast<FATFS const&>(Inode::fs()); }
+
+private:
+    FATInode(FATFS&, FATEntry, NonnullOwnPtr<KString> filename);
+
+    static constexpr u32 root_inode_index = 1;
+
+    static constexpr u32 no_more_clusters = 0x0FFFFFF8;
+    static constexpr u32 cluster_number_mask = 0x0FFFFFFF;
+
+    static constexpr u8 end_entry_byte = 0x00;
+    static constexpr u8 unused_entry_byte = 0xE5;
+
+    static constexpr u8 lfn_entry_text_termination = 0xFF;
+
+    static constexpr u16 first_fat_year_offset = 10; // 1980 (first possible year that can be stored as a fat date) - 1970 (first unix year)
```

Comment:
Comment should be above the code that it comments, not behind it (unless the comment is _very_ short, which it isn't).